### PR TITLE
Rewrite exception callback in C++

### PR DESCRIPTION
**Pull request by vabold**
Notably, this removes the automated scrolling and adds diagonal inputs.

### CODE REVIEW

**Comment by vabold** [comment]
With added GC and classic controller support, this should be good for review.

File: payload/egg/util/eggException.cc [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+#include <common/Clock.hh>
+
+#include "egg/core/eggThread.hh"
+
+extern "C" {
+#include <revolution/kpad.h>
+#include <revolution/os.h>
+#include <revolution/pad.h>
+#include <revolution/vi.h>
+}
```
Comment (stblr) [inline payload/egg/util/eggException.cc:12]:
```suggestion
#include "egg/core/eggThread.hh"

#include <common/Clock.hh>
extern "C" {
#include <revolution/kpad.h>
#include <revolution/os.h>
#include <revolution/pad.h>
#include <revolution/vi.h>
}
```

File: payload/egg/util/eggException.cc [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+    s32 totalLines = nw4r::db::Console_GetTotalLines(console);
+
+    console->m_visible = true;
+    console->m_currentTopLine = lineCount;
+    nw4r::db::Console_DrawDirect(console);
+
+    KPADStatus wStatus;
+    PADStatus gcStatus[4];
+    WPADCLStatus clStatus;
+    u32 input;
```
Comment (stblr) [inline payload/egg/util/eggException.cc:51]: No reason to create these variables so early, as they are not used in the outer scope.

File: payload/egg/util/eggException.cc [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+
+    while (true) {
+        KPADRead(0, &wStatus, 1);
+        PADRead(gcStatus);
+        PADClampCircle(gcStatus);
+        if (classic) {
+            KPADGetUnifiedWpadStatus(0, &clStatus, 1);
+        }
+
+        input = wStatus.buttons & 15;
```
Comment (stblr) [inline payload/egg/util/eggException.cc:64]: In my opinion this is a bit cryptic, I would suggest to have one boolean per direction.

File: payload/revolution/os.h [hunk 4]
```diff
@@ -105,5 +105,6 @@ const char* OSGetAppGamename(void);
 #include "revolution/os/OSThread.h"
 #include "revolution/os/OSTitle.h"
 
-extern u32 OSDisableInterrupts(void);
-extern void OSRestoreInterrupts(u32);
+extern BOOL OSDisableInterrupts(void);
+extern BOOL OSEnableInterrupts();
+extern BOOL OSRestoreInterrupts(BOOL);
```
Comment (stblr) [inline payload/revolution/os.h:110]:
```suggestion
BOOL OSDisableInterrupts(void);
BOOL OSEnableInterrupts();
BOOL OSRestoreInterrupts(BOOL);
```

**Review by stblr — CHANGES_REQUESTED** [review]

File: payload/egg/util/eggException.cc (hunk 1 above)
Comment (vabold) [reply payload/egg/util/eggException.cc]: Applied, though there is a new include to review.

File: payload/egg/util/eggException.cc (hunk 2 above)
Comment (vabold) [reply payload/egg/util/eggException.cc]: Fixed.

File: payload/revolution/os.h (hunk 4 above)
Comment (vabold) [reply payload/revolution/os.h]: Applied.

File: payload/egg/util/eggException.cc (hunk 3 above)
Comment (vabold) [reply payload/egg/util/eggException.cc]: Fixed.

File: payload/egg/util/eggException.hh [hunk 5]
```diff
@@ -0,0 +1,13 @@
+#pragma once
+
+#include <nw4r/db/dbConsole.hh>
+
+namespace EGG {
+
+static constexpr bool CheckGCStickThreshold(s8 stick);
+
+static constexpr bool CheckCLStickThreshold(s16 stick);
```
Comment (stblr) [inline payload/egg/util/eggException.hh:9]:
```suggestion
```

File: payload/egg/util/eggException.cc [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+
+    u32 controllerType;
+    s32 err = WPADProbe(0, &controllerType);
+    bool classic = (err == 0 && controllerType == 2);
+
+    while (true) {
+        KPADStatus wStatus;
+        memset(&wStatus, 0, sizeof(KPADStatus));
+        PADStatus gcStatus[4];
+        WPADCLStatus clStatus;
```
Comment (stblr) [inline payload/egg/util/eggException.cc:57]:
```suggestion
        KPADStatus wStatus{};
        PADStatus gcStatus[4]{};
        WPADCLStatus clStatus{};
```

**Review by stblr — CHANGES_REQUESTED** [review]

File: payload/egg/util/eggException.hh (hunk 5 above)
Comment (vabold) [reply payload/egg/util/eggException.hh]: Applied.

File: payload/egg/util/eggException.cc (hunk 6 above)
Comment (vabold) [reply payload/egg/util/eggException.cc]: Applied.

**Review by stblr — APPROVED** [review]

**Pull request merged**